olp-cpp-sdk-dataservice-read/src/repositories/CatalogRepository.cpp

```diff
@@ -184,6 +184,96 @@ CatalogVersionResponse CatalogRepository::GetLatestVersion(
   return version_response;
 }
 
+client::CancellationToken CatalogRepository::GetLatestVersion(
+    const CatalogVersionRequest& request, CatalogVersionCallback callback) {
+  repository::CatalogCacheRepository repository(
+      catalog_, settings_.cache, settings_.default_cache_expiration);
+
+  const auto fetch_option = request.GetFetchOption();
+  // in case get version online was never called and version was not found in
+  // the cache.
+  CatalogVersionResponse version_response = {
+      {client::ErrorCode::NotFound, "Failed to find a catalog version."}};
+
+  auto func = [=, &request,
+               &version_response]() mutable -> CatalogVersionResponse {
+    auto cached_version = repository.GetVersion();
+
+    // Using `GetStartVersion` to set up a new latest version for CacheOnly
+    // requests in case of absence of the previous latest version or it's less
+    // than the new user set version.
+    if (fetch_option == CacheOnly) {
+      constexpr auto kDefaultStartVersion = 0;
+
+      auto set_version = request.GetStartVersion();
+      if (set_version >= kDefaultStartVersion) {
+        if (!cached_version || set_version > cached_version->GetVersion()) {
+          model::VersionResponse new_response;
+          new_response.SetVersion(set_version);
+          version_response = new_response;
+        }
+      }
+    }
+
+    if (version_response.IsSuccessful()) {
+      const auto new_version = version_response.GetResult().GetVersion();
+      // Write or update a version in the cache, updating happens only when the
+      // new version is greater than cached.
+      if (!cached_version || (*cached_version).GetVersion() < new_version) {
+        repository.PutVersion(version_response.GetResult());
+        if (fetch_option == CacheOnly) {
+          OLP_SDK_LOG_DEBUG_F(
+              kLogTag, "Latest user set version, hrn='%s', version=%" PRId64,
+              catalog_.ToCatalogHRNString().c_str(), new_version);
+        } else {
+          OLP_SDK_LOG_DEBUG_F(
+              kLogTag, "Latest online version, hrn='%s', version=%" PRId64,
+              catalog_.ToCatalogHRNString().c_str(), new_version);
+        }
+      }
+      return version_response;
+    }
+
+    if (cached_version) {
+      OLP_SDK_LOG_DEBUG_F(kLogTag,
+                          "Latest cached version, hrn='%s', version=%" PRId64,
+                          catalog_.ToCatalogHRNString().c_str(),
+                          (*cached_version).GetVersion());
+      version_response = *std::move(cached_version);
+    }
+    return version_response;
+  };
+
+  if (fetch_option != CacheOnly) {
+    return GetLatestVersionOnline(
+        request.GetBillingTag(), [=](CatalogVersionResponse response) mutable {
+          auto version_response = std::move(response);
```

Comment:
You already have a variable `version_response` in an upper scope!
Please get rid of that one or rename this one.

Why do not use `response` itself?